// react-rocketjump/lib/plugins/listDelete/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('@babel/runtime/helpers/extends');
require('@babel/runtime/helpers/objectWithoutPropertiesLoose');
require('rocketjump-core');
var rj = require('../../rj-5794e05d.js');
require('rxjs');
require('rxjs/operators');
var utils = require('rocketjump-core/utils');
require('@babel/runtime/helpers/inheritsLoose');
var objectPathImmutable = require('object-path-immutable');
require('../../index.cjs.js');
require('react');
require('../../debugger-1b0f24ff.js');
require('hoist-non-react-statics');

var TYPE = 'RJ_LIST_DELETE';

var defaultIdentity = function defaultIdentity(action, listItem) {
  return action.item.id === listItem.id;
};

var rjListDelete = function rjListDelete(config) {
  if (config === void 0) {
    config = {};
  }

  var identity = config.identity || defaultIdentity;
  var path = config.path || 'data';
  return rj.rj({
    actions: function actions() {
      return {
        deleteItem: function deleteItem(item) {
          return {
            type: TYPE,
            item: item
          };
        }
      };
    },
    reducer: function reducer(oldReducer) {
      return function (state, action) {
        if (action.type === TYPE) {
          var paginationPath = path.split('.').map(function (item, i, arr) {
            return i === arr.length - 1 ? 'pagination' : item;
          }).join('.');

          if (state && utils.get(state, paginationPath) && config.warnPagination !== false) {
            // eslint-disable-next-line no-console
            console.warn('It seems you are using this plugin on a paginated list. Remember that this plugin is agnostic wrt pagination, and will break it. To suppress this warning, set warnPagination: false in the config object');
          }

          var list = utils.get(state, path);
          if (list) list = list.filter(function (listItem) {
            return !identity(action, listItem);
          });
          return objectPathImmutable.set(state, path, list);
        }

        return oldReducer(state, action);
      };
    }
  });
};

exports.default = rjListDelete;
