// react-rocketjump/lib/plugins/listInsert/index.es.js
import '@babel/runtime/helpers/esm/extends';
import '@babel/runtime/helpers/esm/objectWithoutPropertiesLoose';
import 'rocketjump-core';
import { r as rj } from '../../rj-8a6e6805.js';
import 'rxjs';
import 'rxjs/operators';
import { get } from 'rocketjump-core/utils';
import '@babel/runtime/helpers/esm/inheritsLoose';
import { set } from 'object-path-immutable';
import '../../index.es.js';
import 'react';
import '../../debugger-ad70456b.js';
import 'hoist-non-react-statics';

var TYPE = 'RJ_LIST_INSERT';

var defaultMerge = function defaultMerge(action, list) {
  return (list || []).concat([action.item]);
};

var rjListInsert = function rjListInsert(config) {
  if (config === void 0) {
    config = {};
  }

  var path = config.path || 'data';
  return rj({
    actions: function actions() {
      return {
        insertItem: function insertItem(item) {
          return {
            type: TYPE,
            item: item
          };
        }
      };
    },
    reducer: function reducer(oldReducer) {
      return function (state, action) {
        if (action.type === TYPE) {
          var paginationPath = path.split('.').map(function (item, i, arr) {
            return i === arr.length - 1 ? 'pagination' : item;
          }).join('.');

          if (state && get(state, paginationPath) && config.warnPagination !== false) {
            console.warn('It seems you are using this plugin on a paginated list. Remember that this plugin is agnostic wrt pagination, and will break it. To suppress this warning, set warnPagination: false in the config object');
          }

          var mergeFunc = config.merge ? config.merge : defaultMerge;
          return set(state, path, mergeFunc(action, get(state, path)));
        }

        return oldReducer(state, action);
      };
    }
  });
};

export default rjListInsert;
