// react-rocketjump/lib/plugins/listInsert/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('@babel/runtime/helpers/extends');
require('@babel/runtime/helpers/objectWithoutPropertiesLoose');
require('rocketjump-core');
var rj = require('../../rj-5794e05d.js');
require('rxjs');
require('rxjs/operators');
var utils = require('rocketjump-core/utils');
require('@babel/runtime/helpers/inheritsLoose');
var objectPathImmutable = require('object-path-immutable');
require('../../index.cjs.js');
require('react');
require('../../debugger-1b0f24ff.js');
require('hoist-non-react-statics');

var TYPE = 'RJ_LIST_INSERT';

var defaultMerge = function defaultMerge(action, list) {
  return (list || []).concat([action.item]);
};

var rjListInsert = function rjListInsert(config) {
  if (config === void 0) {
    config = {};
  }

  var path = config.path || 'data';
  return rj.rj({
    actions: function actions() {
      return {
        insertItem: function insertItem(item) {
          return {
            type: TYPE,
            item: item
          };
        }
      };
    },
    reducer: function reducer(oldReducer) {
      return function (state, action) {
        if (action.type === TYPE) {
          var paginationPath = path.split('.').map(function (item, i, arr) {
            return i === arr.length - 1 ? 'pagination' : item;
          }).join('.');

          if (state && utils.get(state, paginationPath) && config.warnPagination !== false) {
            console.warn('It seems you are using this plugin on a paginated list. Remember that this plugin is agnostic wrt pagination, and will break it. To suppress this warning, set warnPagination: false in the config object');
          }

          var mergeFunc = config.merge ? config.merge : defaultMerge;
          return objectPathImmutable.set(state, path, mergeFunc(action, utils.get(state, path)));
        }

        return oldReducer(state, action);
      };
    }
  });
};

exports.default = rjListInsert;
